api/middleware/check-token.js

```diff
@@ -0,0 +1,50 @@
+const fetch = require('node-fetch');
+const { oAuth: oAuthConfig } = require('../../config');
+
+async function getAccessToken(refreshToken) {
+  const searchParams = {
+    client_id: oAuthConfig.clientID,
+    client_secret: oAuthConfig.clientSecret,
+    grant_type: 'refresh_token',
+    refresh_token: refreshToken
+  };
+  const searchParamsString = new URLSearchParams(searchParams).toString();
+  const response = await fetch(
+    `https://www.googleapis.com/oauth2/v4/token?${searchParamsString}`,
+    { method: 'POST' }
+  );
+  const result = await response.json();
+  if (response.status !== 200) throw new Error(JSON.stringify(result));
+  return result;
+}
+
+module.exports = async (req, res, next) => {
+  if (!req.user || !req.isAuthenticated()) {
+    res.sendStatus(401);
+  } else {
+    const { expiry, refreshToken } = req.user;
+    const now = new Date();
+    if (now.getTime() >= expiry) {
+      try {
+        const { access_token: token, expires_in: expiresIn } = await getAccessToken(refreshToken);
+        const updatedExpiry = now.setSeconds(now.getSeconds() + (expiresIn - 300));
```

Comment:
can you explain this time calculation?

Reply:
So the expires_in property that Google returns gives the amount of time in seconds before the token will expire. In order to give a padding window to seamlessly refresh the token, I'm removing 5 minutes (300 seconds) from the expiry time and saving that as the internal value we use. Why 5 mins? I saw it used in an example haha...